imjoy/helper.py

```diff
@@ -122,48 +122,49 @@ def parseEnv(eng, env, work_dir, default_env_name):
-                    parms = shlex.split(_env)
-                    if "-n" in parms:
-                        env_name = parms[parms.index("-n") + 1]
-                    elif "--name" in parms:
-                        env_name = parms[parms.index("--name") + 1]
-                    else:
-                        env_name = default_env_name
-                        envs[i] = _env.replace(
-                            "conda create", "conda create -n " + env_name
-                        )
-
-                    if "-y" not in parms:
-                        envs[i] = _env.replace("conda create", "conda create -y")
-
-                if "conda env create" in _env:
-                    parms = shlex.split(_env)
-                    if "-f" in parms:
-                        try:
-                            env_file = os.path.join(
-                                work_dir, parms[parms.index("-f") + 1]
+                if type(_env) is str:
+                    if "conda create" in _env:
+                        # if not _env.startswith('conda'):
+                        #     raise Exception('env command must start with conda')
+                        if "python=2" in _env:
+                            is_py2 = True
+                        parms = shlex.split(_env)
+                        if "-n" in parms:
+                            virtual_env_name = parms[parms.index("-n") + 1]
+                        elif "--name" in parms:
+                            virtual_env_name = parms[parms.index("--name") + 1]
+                        else:
+                            virtual_env_name = default_env_name
+                            envs[i] = _env.replace(
+                                "conda create", "conda create -n " + virtual_env_name
                             )
-                            with open(env_file, "r") as stream:
-                                env_config = yaml.load(stream)
-                                assert "name" in env_config
-                                env_name = env_config["name"]
-                        except Exception as exc:
+
+                        if "-y" not in parms:
+                            envs[i] = _env.replace("conda create", "conda create -y")
+
+                    if "conda env create" in _env:
+                        parms = shlex.split(_env)
+                        if "-f" in parms:
+                            try:
+                                env_file = os.path.join(
+                                    work_dir, parms[parms.index("-f") + 1]
+                                )
+                                with open(env_file, "r") as stream:
+                                    env_config = yaml.load(stream)
+                                    assert "name" in env_config
+                                    virtual_env_name = env_config["name"]
+                            except Exception as exc:
+                                raise Exception(
```

Comment:
We should just `raise` to keep the context of the exception I think.

Reply:
What if we want to customize the error message? This error will eventually show on the interface , so I try to add some additional hint on what happening.

Reply:
We can use `raise from`.

https://franklingu.github.io/programming/2016/06/30/properly-reraise-exception-in-python/

Sidenote: I think we should start using imjoy specific exceptions when we want to raise exceptions, ie define our own exception classes.

Reply:
Good to know, will try to use that! Yes, we should define our own.